includes/Modules/Tag_Manager.php

```diff
@@ -505,12 +505,14 @@ protected function setup_assets() {
 				'googlesitekit-modules-tagmanager',
 				array(
 					'src'          => $base_url . 'js/googlesitekit-modules-tagmanager.js',
+					// Note that the Tag Manager bundle will make use of the Analytics bundle
+					// if it's available, but can also function without it, hence the Analytics
+					// bundle not appearing in the list of dependencies here.
 					'dependencies' => array(
 						'googlesitekit-api',
 						'googlesitekit-data',
 						'googlesitekit-datastore-site',
 						'googlesitekit-modules',
-						'googlesitekit-modules-analytics',
```

Comment:
I think we should still probably add this conditionally if it is available. That way the code paths that run when it is available still run as today in terms of when the GA bundle is loaded. What do you think?

Reply:
That sounds sensible. I have made this change, but in order to do so we need to allow the Tag Manager module to check whether the Analytics module is available - at present a `Module` has no real awareness of other `Modules`.

The approach I've taken is to pass in the list of available modules to the `get_assets` and `setup_assets` methods. However, it's arguably a bit brittle amending this shared method signature in this way just for the one instance of it. You can see the changes [here](https://github.com/google/site-kit-wp/pull/5713/files/1f73cf7b7bafc3acd6e80557c7660832348f0f79..baabbe5d7923279573bab3be802836afb45e9b3e).

I'm wondering if we should take a different approach:

1. Pass in the entire `Modules` instance to `get_assets` / `setup_assets` for more flexibility and call `$modules->module_exists( 'analytics' )` within `setup_assets` to check if Analytics is present.
2. Amend the `Module` class to hold a reference to the `Modules` instance, passing it into each `Module` when it's instantiated in `Modules::get_available_modules`. Call `$this->modules->module_exists( 'analytics' )` in `setup_assets`. This doesn't feel like a nice solution because it introduces a circular reference between `Modules` and each `Module`.
3. Use a filter instead, in a similar way to how the `googlesitekit_analytics_adsense_linked` filter is used to allow the presence of `AdSense` to affect the `Analytics` module.
4. Something else entirely.

Please take a look, interested to hear your thoughts on this.

Reply:
Thanks – I agree, it would be better to find another way to avoid passing the list through to all those methods. After thinking about it a bit, I think using a filter would be simplest. Basically, the asset handle used for the dependency here would be filtered in by the Analytics module class. This way the handle would only be added if Analytics were available as only available modules are registered. I wouldn't make the whole list of dependencies filterable though, just the one for Analytics, which could default to `false` or anything else falsey such that we could filter it out to not register a non-existent asset as a dependency which would obviously break things.

One idea I was toying with here as a kind of asset-polymorphism would be to register a `no-op` script handle – this could be done using a `Script_Data` instance with no dependencies of its own, or any other args really. Essentially an empty handle. Then here we could do something like this:

```php
array(
	'googlesitekit-api',
	'googlesitekit-data',
	'googlesitekit-datastore-site',
	'googlesitekit-modules',
	'googlesitekit-vendor',
	apply_filters( 'googlesitekit-conditional-analytics-asset', 'googlesitekit-no-op' ),
)
```

That way, if Analytics were disabled, it would use the no-op asset which would technically be "enqueued" but otherwise do nothing. I'm not sure if this would really be better than using a filter without the no-op and just filtering out any falsey values from the list but just an idea.

Reply:
Thanks @aaemnnosttv, I have reverted the changes to the signature for `get_assets` / `setup_assets` and used a filter instead. However I have taken a different approach to the filter than the one you proposed, I hope you agree it makes sense.

I noticed we already have this `googlesitekit_is_module_recoverable` filter in `Modules`:

https://github.com/google/site-kit-wp/blob/6b407087c078a557680f3a783cbc2ba5cbb04344/includes/Core/Modules/Modules.php#L391-L398

This is used as the basis of the `Module::is_recoverable` method:

https://github.com/google/site-kit-wp/blob/6b407087c078a557680f3a783cbc2ba5cbb04344/includes/Core/Modules/Module.php#L875-L884

I figure, especially as we already have this approach, providing a `googlesitekit_module_exists` filter would be generally useful and in keeping with the existing codebase. So I have added this to `Modules` and used the filter directly within `Tag_Manager::setup_assets`. Maybe we should add a `Module::module_exists( $slug )` method to encapsulate the filter usage, but I am not entirely sure it's needed.

Anyhow, the change is [here](https://github.com/google/site-kit-wp/pull/5713/files/701bbc18de22e63308fd11efba4d0fb1dfa31aa6..6b407087c078a557680f3a783cbc2ba5cbb04344), please see what you think.

Reply:
Thanks, this sounds like a good solution to me 👍 Nice work!